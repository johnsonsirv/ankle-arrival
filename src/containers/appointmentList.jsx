import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Spinner from 'react-spinkit';
import * as dispatchActions from '../actions';
import AppointmentDetails from '../components/appointmentDetails';
import { CurrentUserType } from '../utils/prop-types';

const mapStateToProps = state => state;

export const AppointmentList = props => {
  const {
    fetchAppointments,
    appointments: { appointments: appointmentData, isFetching },
    currentUser,
  } = props;

  useEffect(() => {
    fetchAppointments(currentUser.token);
  }, [fetchAppointments, currentUser]);

  return (
    <>
      {isFetching && <Spinner name="three-bounce" fadeIn="none" />}
      <div className="appointmentList">
        {appointmentData.map(appointment => (
          <AppointmentDetails
            appointment={appointment}
            key={appointment.id}
          />
        ))}
      </div>
    </>
  );
};

AppointmentList.propTypes = {
  appointments: PropTypes.shape({
    appointments: PropTypes.arrayOf(
      PropTypes.shape({
        dateOfAppointment: PropTypes.string,
        timeOfAppointment: PropTypes.string,
        description: PropTypes.string,
        doctor_firstname: PropTypes.string,
        doctor_lastname: PropTypes.string,
        id: PropTypes.number,
      }).isRequired,
    ).isRequired,
    isFetching: PropTypes.bool.isRequired,
  }).isRequired,
  fetchAppointments: PropTypes.func.isRequired,
  currentUser: PropTypes.shape(CurrentUserType).isRequired,
};

export default connect(mapStateToProps, dispatchActions)(AppointmentList);
