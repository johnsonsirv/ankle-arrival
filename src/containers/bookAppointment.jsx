import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Joi from 'joi-browser';
import { toast } from 'react-toastify';
import moment from 'moment';
import DateInput from '../components/forms/inputs/dateInput';
import TimeInput from '../components/forms/inputs/timeInput';
import RichTextField from '../components/forms/inputs/richTextField';
import ReadOnlyTextField from '../components/forms/inputs/readOnlyTextField';
import * as dispatchActions from '../actions';

export const BookAppointment = props => {
  const {
    currentUser: { firstname, lastname },
    currentUser,
    doctor,
    doctor: { firstname: doctorFirstname, lastname: doctorLastname },
  } = props;

  const [booking, setBooking] = useState({
    dateOfAppointment: null,
    timeOfAppointment: null,
    description: '',
    currentUser,
    doctor,
    isValid: false,
  });

  const schema = {
    description: Joi.string()
      .required()
      .label('Description'),
    dateOfAppointment: Joi.string()
      .required()
      .label('Date of Appointment'),
    timeOfAppointment: Joi.string()
      .required()
      .label('Time of Appointment'),
  };

  const validateProperty = ({ name, value }) => {
    const obj = { [name]: value };
    const subSchema = { [name]: schema[name] };
    const { error } = Joi.validate(obj, subSchema);
    return error ? error.details[0].message : null;
  };

  const handleSubmit = e => {
    e.preventDefault();
    const {
      dateOfAppointment,
      timeOfAppointment,
      description,
      doctor,
      currentUser,
    } = booking;
    props.addNewAppointment({
      dateOfAppointment,
      timeOfAppointment,
      description,
      currentUser,
      doctor,
    });

    // redirect here
  };

  const handleChange = e => {
    const { value, name } = e.target;
    booking[name] = value;
    const error = validateProperty(e.target);
    booking.isValid = !error;

    setBooking({ ...booking });
    // toastify
    toast.error(error);
  };

  useEffect(() => {
    setBooking({
      currentUser,
      doctor,
    });
  }, [doctor, currentUser]);

  return (
    <div>
      <form>
        <ReadOnlyTextField
          name="doctor"
          value={`${doctorFirstname} ${doctorLastname}`}
        />
        <ReadOnlyTextField
          name="current-user"
          value={`${firstname} ${lastname}`}
        />
        <RichTextField
          name="description"
          value={booking.description}
          onChange={handleChange}
        />
        <DateInput
          onChange={handleChange}
          name="dateOfAppointment"
          value={booking.dateOfAppointment}
        />
        <TimeInput
          onChange={handleChange}
          name="timeOfAppointment"
          value={booking.timeOfAppointment}
        />
        <button
          type="button"
          onClick={handleSubmit}
          id="book-appointment"
          disabled={!booking.isValid}
        >
          Save Appointment
        </button>
      </form>
    </div>
  );
};

BookAppointment.propTypes = {
  doctor: PropTypes.shape({
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    city: PropTypes.string,
    email: PropTypes.string,
    username: PropTypes.string,
    id: PropTypes.number,
  }).isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    token: PropTypes.string,
    firstname: PropTypes.string,
    lastname: PropTypes.string,
  }).isRequired,
  addNewAppointment: PropTypes.func.isRequired,
};

export default connect(null, dispatchActions)(BookAppointment);
