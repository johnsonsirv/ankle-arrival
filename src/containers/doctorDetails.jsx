import React from 'react';

const DotorDetails = () => {
  return <div>doctor details page</div>;
};

export default DotorDetails;
