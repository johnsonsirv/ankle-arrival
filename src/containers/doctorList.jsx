import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Spinner from 'react-spinkit';
import Doctor from '../components/doctor';
import * as dispatchActions from '../actions';

const mapStateToProps = state => state;

export const DoctorList = props => {
  const {
    currentUser,
    doctors: { doctors: doctorData, isFetching },
    fetchDoctors,
  } = props;

  useEffect(() => {
    fetchDoctors(currentUser.token);
  }, [fetchDoctors, currentUser.token]);

  return (
    <>
      {isFetching && <Spinner name="three-bounce" fadeIn="none" />}
      {doctorData.map(doctor => (
        <Doctor doctor={doctor} currentUser={currentUser} key={doctor.id} />
      ))}
    </>
  );
};

DoctorList.propTypes = {
  doctors: PropTypes.shape({
    doctors: PropTypes.arrayOf(
      PropTypes.shape({
        firstname: PropTypes.string,
        lastname: PropTypes.string,
        city: PropTypes.string,
        email: PropTypes.string,
        username: PropTypes.string,
        id: PropTypes.number,
      }).isRequired
    ).isRequired,
    isFetching: PropTypes.bool.isRequired,
  }).isRequired,

  fetchDoctors: PropTypes.func.isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    token: PropTypes.string,
  }).isRequired,
};

export default connect(mapStateToProps, dispatchActions)(DoctorList);
