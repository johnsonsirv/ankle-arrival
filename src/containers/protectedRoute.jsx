/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Route, Redirect } from 'react-router-dom';
import * as dispatchActions from '../actions';

const mapStateToProps = state => state;

const ProtectedRoute = ({
  getCurrentUser,
  isAuthenticated,
  component: Component,
  render,
  ...rest
}) => {
  useEffect(() => {
    getCurrentUser();
  }, []);
  return (
    <Route
      {...rest}
      render={props => {
        if (!isAuthenticated) return <Redirect to="/login" />;
        return Component ? <Component {...props} /> : render(props);
      }}
    />
  );
};

ProtectedRoute.propTypes = {
  getCurrentUser: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps, dispatchActions)(ProtectedRoute);
