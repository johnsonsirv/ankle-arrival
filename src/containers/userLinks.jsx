import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import * as dispatchActions from '../actions';
import ProtectedLinks from '../components/protectedLinks';
import { CurrentUserType } from '../utils/prop-types';

const mapStateToProps = state => state;

export const UserLinks = props => {
  const {
    getCurrentUser,
    currentUser: { isAuthenticated },
  } = props;

  const handleLogout = () => {
    localStorage.removeItem('currentUser');
    window.location = '/';
  };

  useEffect(() => {
    getCurrentUser();
  }, [getCurrentUser]);

  return (
    <>
      <div className="userLinks">
        {!isAuthenticated && (
          <div className="unProtectedLinks">
            <NavLink to="/login">Login</NavLink>
            <NavLink to="/signup">Sign Up</NavLink>
          </div>
        )}
        {isAuthenticated && (
          <div className="protectedLinks">
            <ProtectedLinks />
            <NavLink to="" onClick={handleLogout}>Logout</NavLink>
          </div>
        )}
      </div>
    </>
  );
};

UserLinks.propTypes = {
  getCurrentUser: PropTypes.func.isRequired,
  currentUser: PropTypes.shape(CurrentUserType).isRequired,
};

export default connect(mapStateToProps, dispatchActions)(UserLinks);
