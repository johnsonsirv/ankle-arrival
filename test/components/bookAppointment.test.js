/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable no-undef */
import moment from 'moment';
import { BookAppointment } from '../../src/containers/bookAppointment';

function setup() {
  const props = {
    doctor: {
      id: 1,
      firstname: 'Larry',
      lastname: 'Page',
      email: 'example@larry.com',
      city: 'NY',
      username: 'larryp123',
    },
    currentUser: {
      id: 1,
      username: 'testUser',
      token: 'xxx.yyy.zzz',
      firstname: 'Victor',
      lastname: 'Johnson',
    },
    addNewAppointment: jest.fn(),
  };

  const enzymeWrapper = shallow(<BookAppointment {...props} />);

  return { props, enzymeWrapper };
}

describe('<BookAppointment /> rendering', () => {
  const { props, enzymeWrapper } = setup();

  it('should render correctly', () => {
    expect(enzymeWrapper).toMatchSnapshot();
  });
  it('should receive doctor props', () => {
    const doctor = `${props.doctor.firstname} ${props.doctor.lastname}`;
    expect(enzymeWrapper.find('ReadOnlyTextField').get(0).props.value).toEqual(
      doctor
    );
  });
  it('should receive currentUser props', () => {
    const currentUser = `${props.currentUser.firstname} ${props.currentUser.lastname}`;
    expect(enzymeWrapper.find('ReadOnlyTextField').get(1).props.value).toEqual(
      currentUser
    );
  });
  it('should render 1 <Button>', () => {
    expect(enzymeWrapper.find('button')).toHaveLength(1);
  });
  it('should render 2 <ReadOnlyTextField />', () => {
    expect(enzymeWrapper.find('ReadOnlyTextField')).toHaveLength(2);
  });
  it('should render a <DateInput /> component', () => {
    expect(enzymeWrapper.find('DateInput')).toHaveLength(1);
  });
  it('should render a <TimeInput /> component', () => {
    expect(enzymeWrapper.find('TimeInput')).toHaveLength(1);
  });
  it('should render a <RichTextField /> component', () => {
    expect(enzymeWrapper.find('RichTextField')).toHaveLength(1);
  });
});

describe('<BookAppointment /> interactions', () => {
  const { props, enzymeWrapper } = setup();

  it('should enable <button> if inputs are valid', () => {
    const dateInput = enzymeWrapper.find('DateInput');
    dateInput.simulate('change', {
      target: {
        value: moment().format('2020-03-06'),
        name: 'dateOfAppointment',
      },
    });
    const timeInput = enzymeWrapper.find('TimeInput');
    timeInput.simulate('change', {
      target: { value: '12:00 PM', name: 'timeOfAppointment' },
    });
    const richTextField = enzymeWrapper.find('RichTextField');
    richTextField.simulate('change', {
      target: { value: 'I need to see a doctor', name: 'description' },
    });
    expect(enzymeWrapper.find('#book-appointment').prop('disabled')).toBe(
      false
    );
  });
  it('should disable <button> for click if inputs are invalid', () => {
    const dateInput = enzymeWrapper.find('DateInput');
    dateInput.simulate('change', {
      target: { value: null, name: 'dateOfAppointment' },
    });
    expect(enzymeWrapper.find('#book-appointment').prop('disabled')).toBe(true);
  });
  it('<DateInput /> should show selected date onChange', () => {
    const dateInput = enzymeWrapper.find('DateInput');
    dateInput.simulate('change', {
      target: {
        value: moment().format('2020-03-06'),
        name: 'dateOfAppointment',
      },
    });
    expect(enzymeWrapper.find('DateInput').prop('value')).toEqual('2020-03-06');
  });
  it('<TimeInput /> should show selected time onChange', () => {
    const timeInput = enzymeWrapper.find('TimeInput');
    timeInput.simulate('change', {
      target: { value: '12:00 PM', name: 'timeOfAppointment' },
    });
    expect(enzymeWrapper.find('TimeInput').prop('value')).toEqual('12:00 PM');
  });
  it('<RichTextField /> should show description onChange', () => {
    const richTextField = enzymeWrapper.find('RichTextField');
    richTextField.simulate('change', {
      target: { value: 'I need to see a doctor', name: 'description' },
    });
    expect(enzymeWrapper.find('RichTextField').get(0).props.value).toMatch(
      /I need to see a doctor/
    );
  });
  it('should dispatch saveAppointment action on handleSubmit', () => {
    const dateInput = enzymeWrapper.find('DateInput');
    dateInput.simulate('change', {
      target: {
        value: moment().format('2020-03-06'),
        name: 'dateOfAppointment',
      },
    });
    const timeInput = enzymeWrapper.find('TimeInput');
    timeInput.simulate('change', {
      target: { value: '12:00 PM', name: 'timeOfAppointment' },
    });
    const richTextField = enzymeWrapper.find('RichTextField');
    richTextField.simulate('change', {
      target: { value: 'I need to see a doctor', name: 'description' },
    });

    enzymeWrapper
      .find('#book-appointment')
      .simulate('click', { target: {}, preventDefault: jest.fn() });
    expect(props.addNewAppointment).toHaveBeenCalled();
  });
});
