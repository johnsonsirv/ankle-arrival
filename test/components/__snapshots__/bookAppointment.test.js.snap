// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookAppointment /> rendering should render correctly 1`] = `ShallowWrapper {}`;
